lua/autorun/client/cl_devtools.lua

```diff
@@ -1,144 +1,165 @@
+
+]]--
+
+-- This file is subject to copyright - contact swampservers@gmail.com for more information.
 -- INSTALL: CINEMA
 SWAMP_DEV = SWAMP_DEV or {}
-SWAMP_DEV.isDev = SWAMP_DEV.isDev
+-- Are we in development mode?
+SWAMP_DEV.isDev = false
+-- The delay on checking if we need to refresh files
 SWAMP_DEV.refreshDelay = 2
-SWAMP_DEV.itemStruct = SWAMP_DEV.itemStruct
+
+local currentItemStruct = nil
+
+--- This stores file times for certain root folder.
+--- FileNames[key] -> SaveTime[value]
 SWAMP_DEV.fileTimes = SWAMP_DEV.fileTimes or {}
+ 
+-- lua/STRUCT/*
 
-SWAMP_DEV.structTypes = {
+SWAMP_DEV.structEnvironments = {
     ["weapons"] = function(curClass)
-        SWEP = weapons.GetStored(curClass)
+        local found = weapons.GetStored(curClass)
+        if not found then return end
+        return {
+            SWEP = found
+        }
     end,
     ["entities"] = function(curClass)
-        ENT = scripted_ents.GetStored(curClass)
+        local found = scripted_ents.GetStored(curClass)
+        if not found then return end
+       
+        return {
+            ENT = found
+        }
     end,
     ["effects"] = function(curClass)
-        local fxTab = effects.GetList()
-        local fx = "effects/" .. curClass
-
-        for _, v in ipairs(fxTab) do
-            if (v.Folder == fx) then
-                EFFECT = v
-
-                return
+        local allEffects = effects.GetList()
+        local targetEffect = "effects/" .. curClass
+
+        for _, effect in ipairs(allEffects) do
+            if (effect.Folder == targetEffect) then
+                return {
+                    EFFECT = effect
+                }
             end
         end
     end
 }
 
--- Returns a string to print
-local function refreshFile(path, filename)
-    local code = file.Read(path, "MOD")
-
-    -- current filename as class format
-    if SWAMP_DEV.itemStruct then
-        local parentFolder = string.gsub(string.sub(path, 1, -#filename - 2), ".*/", "")
 
-        -- if parent isn't a struct then use parent as class, else use filename
-        if (not SWAMP_DEV.structTypes[parentFolder]) then
-            SWAMP_DEV.structTypes[SWAMP_DEV.itemStruct](parentFolder)
-        else
-            SWAMP_DEV.structTypes[SWAMP_DEV.itemStruct](string.TrimRight(string.gsub(path, "^.*/", ""), ".lua"))
+local rootDirectory = "addons/contrib/"
+local refreshFolder
+local loadFile
+do 
+    local loadFunc = function() end
+    function loadFile(filePath, environmentVars)
+        local code = file.Read(filePath, "MOD")
+        if not code then error("No file exists with the file path: " .. filePath) end
+        environmentVars = environmentVars or {}
+        loadFunc = CompileString(code, "SwapDevTools/" .. filePath)
+        -- setup our environment
+        local environment = _G
+        for key,var in pairs(environmentVars) do
+            environment[key] = var
         end
-    end
+        
+        debug.setfenv(loadFunc,environment)
```

Comment:
Aren't you just setting _G? I don't see why this is necessary

Reply:
No, it's not just setting _G, as we're loading the environment of the file, the idea is that in the future this could be used to sandbox the code being reloaded to an extent.

Reply:
How is that potentially useful? The point of this code is the ability to test clientside changes to contrib on the live server. The committed code doesn't run with any special environment, so, why would you want to do that?

Reply:
It was for a potential feature. I can remove it, if you'd like that's fine.

Reply:
Yeah i would say remove it, I am curious what the usefulness of that is though

Reply:
The intended feature of that is so that you could add a certain layer of sandbox-ing around reloaded scripts, so if you didn't want certain reloaded scripts having access to certain global functions that's the place to do it. This was prior to my understanding of you wanting a hard coded list of allowed users.